File: typescript/Creational/04-Prototype/Product/JobCard.ts
const shortId = () => Math.random().toString(36).slice(2, 10);

export interface IPrototype<T> {
  clone(): T;
}

export class JobCard implements IPrototype<JobCard> {
  readonly jobId: string = shortId();
  title: string;
  readonly category: string;
  priority: string;
  scheduledAt?: Date;
  readonly parts: string[];
  readonly photoUrls: string[];
  specialInstructions?: string;
  readonly requiresInspection: boolean;
  readonly hasWarranty: boolean;
  unitReference?: string;

  constructor(config: {
    title: string;
    category: string;
    priority: string;
    scheduledAt?: Date;
    parts: string[];
    photoUrls: string[];
    specialInstructions?: string;
    requiresInspection: boolean;
    hasWarranty: boolean;
    unitReference?: string;
  }) {
    this.title = config.title;
    this.category = config.category;
    this.priority = config.priority;
    this.scheduledAt = config.scheduledAt;
    this.parts = config.parts;
    this.photoUrls = config.photoUrls;
    this.specialInstructions = config.specialInstructions;
    this.requiresInspection = config.requiresInspection;
    this.hasWarranty = config.hasWarranty;
    this.unitReference = config.unitReference;
  }

  // Deep clone — new jobId, copied arrays, Date copied by value
  clone(): JobCard {
    return new JobCard({
      title: this.title,
      category: this.category,
      priority: this.priority,
      scheduledAt: this.scheduledAt ? new Date(this.scheduledAt) : undefined,
      parts: [...this.parts],
      photoUrls: [...this.photoUrls],
      specialInstructions: this.specialInstructions,
      requiresInspection: this.requiresInspection,
      hasWarranty: this.hasWarranty,
      unitReference: this.unitReference,
    });
  }

  printSummary(): void {
    console.log(`\n=== Job Card #${this.jobId} ===`);
    console.log(`  Title    : ${this.title}`);
    console.log(`  Category : ${this.category}`);
    console.log(`  Priority : ${this.priority}`);
    console.log(`  Unit     : ${this.unitReference ?? "N/A"}`);

    if (this.scheduledAt)
      console.log(`  Scheduled: ${this.scheduledAt.toLocaleString()}`);

    if (this.parts.length) console.log(`  Parts    : ${this.parts.join(", ")}`);

    if (this.specialInstructions)
      console.log(`  Notes    : ${this.specialInstructions}`);

    console.log(`  Inspection Required : ${this.requiresInspection}`);
    console.log(`  Warranty            : ${this.hasWarranty}`);
  }
}

File: csharp/Creational/04-Prototype/Product/JobCard.cs
public class JobCard : IPrototype<JobCard>
{
    public string JobId { get; init; } = Guid.NewGuid().ToString()[..8];
    public string Title { get; set; } = string.Empty;
    public string Category { get; init; } = string.Empty;
    public string Priority { get; set; } = "NORMAL";
    public DateTime? ScheduledAt { get; set; }
    public List<string> Parts { get; init; } = new();
    public List<string> PhotoUrls { get; init; } = new();
    public string? SpecialInstructions { get; set; }
    public bool RequiresInspection { get; init; }
    public bool HasWarranty { get; init; }
    public string? UnitReference { get; set; }

    // Deep clone — new JobId, new lists, same values
    public JobCard Clone() => new JobCard
    {
        // JobId intentionally regenerated — each clone is a distinct job
        Title = Title,
        Category = Category,
        Priority = Priority,
        ScheduledAt = ScheduledAt,
        Parts = new List<string>(Parts),
        PhotoUrls = new List<string>(PhotoUrls),
        SpecialInstructions = SpecialInstructions,
        RequiresInspection = RequiresInspection,
        HasWarranty = HasWarranty,
        UnitReference = UnitReference,
    };

    public void PrintSummary()
    {
        Console.WriteLine($"\n=== Job Card #{JobId} ===");
        Console.WriteLine($"  Title    : {Title}");
        Console.WriteLine($"  Category : {Category}");
        Console.WriteLine($"  Priority : {Priority}");
        Console.WriteLine($"  Unit     : {UnitReference ?? "N/A"}");

        if (ScheduledAt.HasValue)
            Console.WriteLine($"  Scheduled: {ScheduledAt:dd MMM yyyy HH:mm}");

        if (Parts.Any())
            Console.WriteLine($"  Parts    : {string.Join(", ", Parts)}");

        if (!string.IsNullOrEmpty(SpecialInstructions))
            Console.WriteLine($"  Notes    : {SpecialInstructions}");

        Console.WriteLine($"  Inspection Required : {RequiresInspection}");
        Console.WriteLine($"  Warranty            : {HasWarranty}");
    }
}
